ambry-store/src/main/java/com.github.ambry.store/IndexValue.java

```diff
@@ -13,110 +13,343 @@
+    }
+  }
+
+  /**
+   * @return the operation time of the index value
+   */
+  int getOperationTimeInSecs() {
+    if (version == 0) {
+      return -1;
+    } else {
+      return value.getInt(OPERATION_TIME_RELATIVE_OFFSET);
+    }
   }
 
+  /**
+   * @return the serviceId of the {@link IndexValue}
+   */
+  short getServiceId() {
+    if (version == 0) {
+      return 0;
+    } else {
+      return value.getShort(SERVICE_ID_RELATIVE_OFFSET);
+    }
+  }
+
+  /**
+   * @return the containerId of the {@link IndexValue}
+   */
+  short getContainerId() {
+    if (version == 0) {
+      return 0;
+    } else {
+      return value.getShort(CONTAINER_ID_RELATIVE_OFFSET);
+    }
+  }
+
+  /**
+   * Sets the {@link Flags} for the {@link IndexValue}
+   * @param flag the {@link Flags} that needs to be set in the {@link IndexValue}
+   */
   void setFlag(Flags flag) {
-    value.put(Blob_Size_In_Bytes + Offset_Size_In_Bytes, (byte) (getFlags() | (1 << flag.ordinal())));
+    value.put(FLAGS_RELATIVE_OFFSET, (byte) (getFlags() | (1 << flag.ordinal())));
   }
 
+  /**
+   * Updates the {@link Offset} of the {@link IndexValue}
+   * @param newOffset the new {@link Offset} to be updated for the {@link IndexValue}
+   */
   void setNewOffset(Offset newOffset) {
     long originalMessageOffset = offset.getName().equals(newOffset.getName()) ? offset.getOffset() : -1;
     offset = newOffset;
-    value.putLong(Blob_Size_In_Bytes, offset.getOffset());
-    value.putLong(Blob_Size_In_Bytes + Offset_Size_In_Bytes + Flag_Size_In_Bytes + Expires_At_Ms_Size_In_Bytes,
-        originalMessageOffset);
+    value.putLong(OFFSET_RELATIVE_OFFSET, offset.getOffset());
+    if (version == 0) {
+      value.putLong(ORIGINAL_MESSAGE_OFFSET_V0_RELATIVE_OFFSET, originalMessageOffset);
+    } else {
+      value.putLong(ORIGINAL_MESSAGE_OFFSET_V1_RELATIVE_OFFSET, originalMessageOffset);
+    }
   }
 
+  /**
+   * Sets the size of the {@link IndexValue}
+   * @param size the size that needs to be set for the {@link IndexValue}
+   */
   void setNewSize(long size) {
     value.putLong(0, size);
   }
 
+  /**
+   * @return the {@link ByteBuffer} representation of this {@link IndexValue}
+   */
   ByteBuffer getBytes() {
     return value;
   }
 
   @Override
   public String toString() {
     return "Offset: " + offset + ", Size: " + getSize() + ", Deleted: " + isFlagSet(Flags.Delete_Index)
-        + ", ExpiresAtMs: " + getExpiresAtMs() + ", Original Message Offset: " + getOriginalMessageOffset();
+        + ", ExpiresAtMs: " + getExpiresAtMs() + ", Original Message Offset: " + getOriginalMessageOffset() + (
+        version == PersistentIndex.VERSION_1 ? ", OperationTimeAtSecs " + getOperationTimeInSecs() + ", ServiceId "
+            + getServiceId() + ", ContainerId " + getContainerId() : "");
+  }
+}
+
+/**
+ * Builder class to assist in contructing the {@link IndexValue}
+ */
+class IndexValueBuilder {
```

Comment:
Already IndexValue had 3 constructors and with new fields added which could have defaults(operaiton time), chances are that callers might get confused. Even I saw that in tests, some constructors were wrongly used and hence the introduction of builder.